File: Days 001 - 010/Day 02/ProductsOfOtherArrayElements.cpp
#include <iostream>
#include <vector>

std::vector<int> GetProductArrayNoDivision(const std::vector<int>& numbers)
{
	const unsigned int SIZE = numbers.size();
	int productCounter = 1;
	std::vector<int> increasingProducts(SIZE);
	
	for (unsigned int i = 0; i < SIZE; i++)
	{
		increasingProducts[i] = productCounter;
		productCounter *= numbers[i];
	}

	productCounter = 1;
	std::vector<int> decreasingProducts(SIZE);

	for (int i = SIZE - 1; i >= 0; i--)
	{
		decreasingProducts[i] = productCounter;
		productCounter *= numbers[i];
	}

	std::vector<int> finalProducts(SIZE);

	for (unsigned int i = 0; i < SIZE; i++)
	{
		finalProducts[i] = increasingProducts[i] * decreasingProducts[i];
	}

	return finalProducts;
}

template <typename T>
void PrintVector(const std::vector<T>& vector)
{
	for (const T& item : vector)
	{
		std::cout << item << " ";
	}

	std::cout << "\n";
}

int main(int argc, char* argv[])
{
	std::vector<int> numbers { 1, 2, 3, 4, 5 };
	std::vector<int> moreNumbers { 3, 2, 1 };

	std::vector<int> numberProducts = GetProductArrayNoDivision(numbers);
	std::vector<int> moreNumberProducts = GetProductArrayNoDivision(moreNumbers);
	
	PrintVector(numberProducts);
	std::cout << "\n";
	PrintVector(moreNumberProducts);

	std::cin.get();

	return 0;
}

File: Days 001 - 010/Day 02/ProductsOfOtherArrayElements.lua
function get_product_array_without_division(array)
	local SIZE = #array
	local product_counter = 1
	local increasing_products = { }

	for i = 1, SIZE
	do
		increasing_products[i] = product_counter
		product_counter = product_counter * array[i]
	end

	product_counter = 1
	local decreasing_products = { }

	for i = SIZE, 1, -1
	do
		decreasing_products[i] = product_counter
		product_counter = product_counter * array[i]
	end

	local combined_products = { }

	for i = 1, SIZE
	do
		combined_products[i] = increasing_products[i] * decreasing_products[i]
	end

	return combined_products
end

function print_array(array)
	for i = 1, #array
	do
		io.write(array[i])

		if i ~= #array
		then
			io.write(", ")
		end
	end

	io.write("\n")
end

print_array(get_product_array_without_division({ 1, 2, 3, 4, 5 }))
print_array(get_product_array_without_division({ 3, 2, 1 }))